### PR TITLE
Mm 45156 RDP initial cypress e2e

### PR DESCRIPTION
**Pull request by trilopin**
#### Summary
E2E tests structure for run details page 

header:
- changed URL for copy link
- e2e tests for participant and viewer
- e2e for viewer skipped until https://github.com/mattermost/mattermost-plugin-playbooks/pull/1259 is merged

Status update:
- disable "View all updates" when there are no updates
- remove the recent status update for viewers if no updates and the status is not in progress
- remove post update and request update buttons if the status is not in progress
- bypass channel permissions to post an update (just need to be a participant)
- e2e tests 

#### Ticket Link
Related https://mattermost.atlassian.net/browse/MM-45156

#### Checklist
- [ ] ~~Telemetry updated~~
- [ ] ~~Gated by experimental feature flag~~
- [x] Unit tests updated

### CODE REVIEW

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+                getHeaderIcon('.icon-product-channels').click();
+
+                // * Assert we navigated correctly
+                cy.url().should('include', `${testTeam.name}/channels/the-run-name`);
+            });
+
+            describe('run actions', () => {
+                describe('modal behaviour', () => {
+                    it('shows and hides as expected', () => {
+                        // should be disabled
```
Comment (gbochora) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:461]: Do you mean that header `Run Actions` button should be disabled/hidden?

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+                cy.findByTestId('run-header-section').find('h1').click();
+
+                getDropdownItemByText('Copy link').click().then(() => {
+                    // * Verify clipboard content
+                    cy.get('@clipboard').its('contents').should('contain', `/playbooks/run_details/${playbookRun.id}`);
+                });
+            });
+
+            describe('finish run', () => {
+                it('can be confirmed', () => {
```
Comment (gbochora) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:489]: Do we plan here to check if the `Finish run` item is missing?

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+                    // do sth
+                });
+            });
+
+            describe('run actions', () => {
+                it('modal can not be opened', () => {
+                    // do sth
```
Comment (gbochora) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:496]: I think(2/5) `Viewer` should be able to open the `Run actions` modal to view configured actions.

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+
+                    // * Assert modal disappeared
+                    cy.findByRole('dialog', {name: /Run Actions/i}).should('not.exist');
+                });
+            });
+        });
+    });
+
+    // waiting for https://github.com/mattermost/mattermost-plugin-playbooks/pull/1259
+    describe.skip('viewer', () => {
```
Comment (gbochora) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:399]: Some of the tests are the same for viewers and participants. Would it be better to reuse code?

File: webapp/src/selectors.ts [hunk 5]
```diff
@@ -68,6 +68,7 @@ export const getRun = (playbookRunId: string, teamId?: string, channelId?: strin
     };
 };
 
+// @deprecated: now we should check playbookRun.participants ang ignore channel members
```
Comment (gbochora) [inline webapp/src/selectors.ts:71]: type:
```suggestion
// @deprecated: now we should check playbookRun.participants and ignore channel members
```

**Review by gbochora — APPROVED** [review]
Great job @trilopin! 🥳  I left some questions and optional suggestions.

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js (hunk 3 above)
Comment (trilopin) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:496]: Not sure about it, as a viewer you are interested more in "what's the progress of the run" than in "How is internally configured?". Could we end up informing about broadcasts on lots of channels we don't have access to (more likely in a viewer than a participant)?

But also 2/5 on that thought.

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+
+    const getHeaderIcon = (selector) => {
+        return cy.findByTestId('run-header-section').find(selector);
+    };
+
+    const getDropdownItemByText = (text) => {
+        return cy.findByTestId('run-header-section').findByTestId('dropdownmenu').findByText(text);
+    };
+
+    describe('participant', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:66]: Nit: I think it's easier to read
```suggestion
    describe('as participant', () => {
```

File: tests-e2e/cypress/integration/runs/rdp_main_statusupdate_spec.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+            ownerUserId: testUser.id,
+        }).then((playbookRun) => {
+            testRun = playbookRun;
+
+            // # Visit the playbook run
+            cy.visit(`/playbooks/run_details/${playbookRun.id}`);
+        });
+    });
+
+    describe('participant', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_statusupdate_spec.js:61]: Same nit:
```suggestion
    describe('as participant', () => {
```

File: tests-e2e/cypress/integration/runs/rdp_main_statusupdate_spec.js [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                cy.getStatusUpdateDialog().findByTestId('modal-confirm-button').click();
+
+                // * Check new due date
+                cy.findByTestId('update-due-date-text').contains('Update due');
+                cy.findByTestId('update-due-date-time').contains('in 15 minutes');
+            });
+        });
+    });
+
+    describe('viewer', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_statusupdate_spec.js:112]: Same nit:
```suggestion
    describe('as viewer', () => {
```

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js (hunk 4 above)
Comment (agarciamontoro) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:399]: +1

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js (hunk 1 above)
Comment (agarciamontoro) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:461]: This was meant to simply check whether the modal shows / hides when clicking on the button / closing it.

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js (hunk 3 above)
Comment (agarciamontoro) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:496]: Channel actions can be viewed by channel members that are not channel admins. I think this should be the same: viewers should be able to see what's configured, but not change it.

File: tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            playbookRunName: 'the run name',
+            ownerUserId: testUser.id,
+        }).then((playbookRun) => {
+            // # Visit the playbook run
+            cy.visit(`/playbooks/run_details/${playbookRun.id}`);
+        });
+    });
+
+    // UnSkip after giorgi's merge
+    it.skip('redirects to not found error if the playbook run is unknown', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js:52]: Should we create tickets for these? Or at least mark them with TODOs? I don't know what changes this refer to, either.

**Review by agarciamontoro — APPROVED** [review]
This is great! Thank you for all the work here, Jose :)

File: tests-e2e/cypress/integration/runs/actions_spec.js [hunk 10]
```diff
@@ -6,6 +6,9 @@
 // - [*] indicates an assertion (e.g. * Check the title)
 // ***************************************************************
 
+// This file could be removed after completing the migration to Run Detail Page
+// The equivalent test is rdp_main_header_spec.js
+
```
[the remark this answers is not in the document]
Comment (trilopin) [reply tests-e2e/cypress/integration/runs/actions_spec.js:11]: I created https://mattermost.atlassian.net/browse/MM-45451 for the cleanup, maybe we should make a list of what to clean inside

File: tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js (hunk 9 above)
Comment (trilopin) [reply tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js:52]: is already done, now it's merged I have to revisit all skipped tests

File: tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        });
+    });
+
+    it('redirects to not found error if the playbook run is unknown', () => {
+        // # Visit the URL of a non-existing playbook run
+        cy.visit('/playbooks/run_details/an_unknown_id');
+
+        // * Verify that the user has been redirected to the playbook runs not found error page
+        cy.url().should('include', '/playbooks/error?type=playbook_runs');
+    });
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js:57]: This test threw me off. It does not belong here, right?

File: tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+            // # Cancel changes
+            cy.findByText('Cancel').click();
+
+            // * Assert that data has changed
+            cy.findByTestId('rendered-text').contains('Add a run summary');
+        });
+
+        // * Assert last edition date is visible
+        cy.findByTestId('run-summary-section').should('not.contain', 'Last edited');
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js:117]: Typos
```suggestion
            // * Assert that data has not changed
            cy.findByTestId('rendered-text').contains('Add a run summary');
        });

        // * Assert last edition date is not visible
        cy.findByTestId('run-summary-section').should('not.contain', 'Last edited');
```

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+            cy.get('#copy-run-link-tooltip').should('contain', 'Copy link to run');
+
+            stubClipboard().as('clipboard');
+            getHeaderIcon('.icon-link-variant').click().then(() => {
+                // * Verify that tooltip text changed
+                cy.get('#copy-run-link-tooltip').should('contain', 'Copied!');
+
+                // * Verify clipboard content
+                cy.get('@clipboard').its('contents').should('contain', `/playbooks/run_details/${playbookRun.id}`);
+            });
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:91]: This `stubClipboard` thing is cool!

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 14]
```diff
@@ -0,0 +1,494 @@
...
+            }).then((run) => {
+                playbookRun = run;
+
+                // # Visit the playbook run
+                cy.visit(`/playbooks/run_details/${playbookRun.id}`);
+            });
+        });
+
+        describe('title and icons', () => {
+            commonHeaderTests();
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:139]: Very nice modularization!

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 15]
```diff
@@ -0,0 +1,494 @@
...
+                    cy.get('#confirmModal').find('h1').contains('Confirm finish run');
+
+                    // # Click on confirm
+                    cy.get('#confirmModal').get('#confirmModalButton').click();
+
+                    // # Open dropdown
+                    cy.findByTestId('run-header-section').find('h1').click();
+
+                    // * Assert option is not anymore in context dropdown
+                    getDropdownItemByText('Finish run').should('not.exist');
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:351]: Nit: I think everywhere we use `getDropdownItemByText`, it is preceded by first opening the dropdown. What if we add the `cy.findByTestId('run-header-section').find('h1').click();` to `getDropdownItemByText` directly?

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 16]
```diff
@@ -0,0 +1,494 @@
...
+
+                it('can be canceled', () => {
+                    // # Open dropdown
+                    cy.findByTestId('run-header-section').find('h1').click();
+
+                    // # Click on finish run
+                    getDropdownItemByText('Finish run').click();
+
+                    // * Check that status badge is in-progress
+                    cy.findByTestId('run-header-section').findByTestId('badge').contains('In Progress');
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:365]: Same comment here about swapping these two steps
```suggestion
                    // * Check that status badge is in-progress
                    cy.findByTestId('run-header-section').findByTestId('badge').contains('In Progress');

                    // # Click on finish run
                    getDropdownItemByText('Finish run').click();
```

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 17]
```diff
@@ -0,0 +1,494 @@
...
+            });
+
+            describe('run actions', () => {
+                describe('modal behaviour', () => {
+                    it('modal can be opened read-only', () => {
+                        // # Open dropdown
+                        cy.findByTestId('run-header-section').find('h1').click();
+
+                        // # Click on finish run
+                        getDropdownItemByText('Run actions').click();
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:444]: Typo
```suggestion
                        // # Click on run actions
                        getDropdownItemByText('Run actions').click();
```

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
This is great, Jose, thank you very much for all this work. I left some comments, but I think once we solve these, we'll be able to merge.

File: webapp/src/components/backstage/playbook_runs/playbook_run/status_update.tsx [hunk 18]
```diff
@@ -165,32 +185,53 @@ export const ParticipantStatusUpdate = ({id, playbookRun, openRHS}: ParticipantP
...
-                        {formatMessage({defaultMessage: 'Post update'})}
-                    </ActionButton>
+                    {playbookRun.current_status === PlaybookRunStatus.InProgress ? (
+                        <ActionButton
+                            data-testid={'post-update-button'}
+                            onClick={postUpdate}
+                        >
+                            {formatMessage({defaultMessage: 'Post update'})}
+                        </ActionButton>
+                    ) : null}
```
[the remark this answers is not in the document]
Comment (trilopin) [reply webapp/src/components/backstage/playbook_runs/playbook_run/status_update.tsx:226]: you need it if you want to use a ternary which I usually prefer to &&

File: tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js (hunk 11 above)
Comment (trilopin) [reply tests-e2e/cypress/integration/runs/rdp_main_summary_spec.js:57]: you're right, I created a new generic file with only that test

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 19]
```diff
@@ -0,0 +1,494 @@
...
+                                    ownerUserId: testUser.id,
+                                });
+                            });
+
+                            // # Navigate to the run page
+                            cy.visit(`/${testTeam.name}/channels/run-with-actions-preconfigured`);
+                            cy.findByRole('button', {name: /Run details/i}).click({force: true});
+
+                            // # Open the run actions modal
+                            openRunActionsModal();
```
[the remark this answers is not in the document]
Comment (trilopin) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:200]: you are right, I make link from title to use the experimental features toggle to navigate to the right url

File: webapp/src/components/backstage/playbook_runs/playbook_run/status_update.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply webapp/src/components/backstage/playbook_runs/playbook_run/status_update.tsx:226]: Ugh, yes, I meant `&&`. I'm ok with the ternary, but I'm just curious: why do you prefer it over `&&`?

File: tests-e2e/cypress/integration/runs/rdp_main_header_spec.js [hunk 20]
```diff
@@ -0,0 +1,468 @@
...
+                        cy.findByRole('dialog', {name: /Run Actions/i}).should('not.exist');
+
+                        // # Open the run actions modal
+                        openRunActionsModal();
+
+                        // * Verify that saving the modal hides it
+                        saveRunActionsModal();
+                    });
+
+                    it.only('honours the settings from the playbook', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:166]: Just one more only! :joy:

Comment (trilopin) [reply tests-e2e/cypress/integration/runs/rdp_main_header_spec.js:166]: omg... sorry

**Pull request merged**